### ISSUE TITLE
CUDA support for default Open-MPI implementation

### ISSUE THREAD

**Issue by tbrandvik** [aws/aws-parallelcluster#1162]
**Environment:**
 - AWS ParallelCluster / CfnCluster version `aws-parallelcluster-2.4.0`
 - OS: `centos7`
 - Scheduler: `slurm`
 - Master instance type: `c5n.large`
 - Compute instance type: `p3.2xlarge`

**Bug description and how to reproduce:**
The default Open-MPI (/opt/amazon/efa) is not compiled with CUDA support. 

Steps to reproduce:

1. Run `ompi_info --parsable --all | grep mpi_built_with_cuda_support:value` on a compute node with a GPU

The output is as follows, indicating that CUDA support is disabled

`mca:mpi:base:param:mpi_built_with_cuda_support:value:false`

Is there a particular reason why CUDA support is not enabled or would it be possible to compile the default Open-MPI version to support CUDA in future releases?

**Comment by sean-smith** [comment]
Thanks for pointing this out, we install openmpi from the efa installers, which in turn build the rpm's. We can add cuda support in a future release, until then you'll need to compile from source:

See https://www.open-mpi.org/faq/?category=buildcuda

**Comment by tbrandvik** [comment]
Perfect - thanks for the quick response!

**Comment by tbrandvik** [comment]
I just wanted to check if CUDA support is enabled in the recent 2.4.1 release?

**Comment by enrico-usai** [comment]
Hi @tbrandvik 
you can see the details of each release in the [Release Notes](https://github.com/aws/aws-parallelcluster/releases).

Unfortunately the CUDA support has not been added yet.

**Comment by afernandezody** [comment]
Hello,
I was beginning to try and build clusters with GPU instances and saw this post, which seems related to an encountered problem while building CUDA-aware OpenMPI. The issue is that even though the compilations work fine, the GPU is not used by MPI jobs. Let me list the steps:
i) Spinning up an instance with the awsparallelcluster CentOS distribution (v2.4.1)
ii) CUDA (v10.0) is installed but unabled, so I enable it by modifying $PATH and $LD_LIBRARY_PATH (via source .bashrc)
iii) gdrcopy and UCX are built  
iv) I build OpenMPI from source as Sean suggested (pretty sure this was done correctly as the UCX team confirmed it in a separate thread)     
Once this is done, MPI jobs run with no error message or similar, but they just ignore the GPU altogether. An example is the OSU microbenchmarks:
```
mpirun -np 2 ./osu_latency 
mpirun -np 2 ./osu_latency D H 
mpirun -np 2 --mca pml ucx --mca btl ^smcuda,openib ./osu_latency D H (the UCX team suggested using this syntax)

```I get very, very similar latencies from the 3 calls and the GPU shows no activity at all.
I have gone over the steps and installation a few times, and my only explanation would be that the virtualized environment is somehow interfering with CUDA-enable MPI but cannot pinpoint the root of the problem. Of course, I'd more than welcome anyone pointing a different explanation. Also, I have two specific questions:
-Has anyone noticed a similar problem trying to build CUDA-aware MPI in an AWS instance/cluster?
-Does the aws-parallelcluster team recommend building CUDA-aware OpenMPI with UCX and gdrcopy support or follow a different route with some other communication framework?
Thanks,
Arturo

**Comment by tbrandvik** [comment]
Hi @enrico-usai, 

I wanted to check if there are any plans to add this in upcoming releases?

Thanks,
Tobias

**Comment by oplatek** [comment]
Hi, is there any update?

**Comment by Srijan1214** [comment]
Hi,
I am kind of new with AWS and was hoping if I could get some help with compiling openmpi with Cuda support from source. As far as I know, I can only build into the head node using AWS batch. How can I build this for all of the compute instances?

**Comment by wzamazon** [comment]
>  As far as I know, I can only build into the head node using AWS batch. How can I build this for all of the compute instances?

The compute instance and the head node have shared disk (such as `/home`). So if you compile open mpi on head node under your home directory, the compute node should have access to that.

**Comment by enrico-usai** [comment]
Resolving this since as part of [3.9.0 release](https://github.com/aws/aws-parallelcluster/releases/tag/v3.9.0) we included EFA installer 1.30.0 that contains both openmpi40-aws-4.1.6-2 and openmpi50-aws-5.0.0-11.

Open MPI 4 from EFA installer does not support CUDA-awareness. Open MPI 5 does.
